x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -552,14 +574,17 @@ static ApiKeyCredentials getCredentialsFromHeader(ThreadContext threadContext) {
     }
 
     // Protected instance method so this can be mocked
-    protected boolean verifyKeyAgainstHash(String apiKeyHash, ApiKeyCredentials credentials) {
+    protected void verifyKeyAgainstHash(String apiKeyHash, ApiKeyCredentials credentials, ActionListener<Boolean> listener) {
         final char[] apiKeyHashChars = apiKeyHash.toCharArray();
```

Comment:
I think this is a subtle but great suggestion. We want minimize the duration a secret stay in memory. When it is outside of the runnable, it could stay in memory for very long if thread pool is busy. Moved it into the runnable as you suggested.